HW3/P4/median_filter.cl

```diff
@@ -1,6 +1,42 @@
 #include "median9.h"
 
-// 3x3 median filter
+static float
+get_clamped_value(__global __read_only float *in_values,
+                  int w, int h,
+                  int x, int y, int halo)
+/*# This function is modified from P5. 
+  # First, check that buffer coordinates are within global bounds.
+  # If not, assign out-of-bounds coordinates the value of nearest valid pixel.
+*/
+{
+    if (x<0) {
+      if (y<0) {
+        return in_values[0];
+      } else if (y>=h) {
+        return in_values[(y-halo) * w + (x + halo)];
+      } else {
+        return in_values[y * w + (x + halo)];
+      }
+    } else if (y<0) {
+      if (x>=w) {
+        return in_values[(y+halo) * w + (x-halo)];
+      } else { //#we already account for x<0&&y<0 in the (x<0) block above
+        return in_values[(y+halo) * w + x];
+      }
+    } else if (x>=w) {
+      if (y>=h) {
+        return in_values[(y-halo) * w + (x-halo)];
+      } else { //#y<0 && x>=w is covered above
+        return in_values[y * w + (x-halo)];
+      }
+    } else if (y>=h) { # all out-of-bounds x cases are covered, we only need the in-bounds x case here
+      return in_values[(y-halo) * w + x];
+    } else {
+      return in_values[y * w + x];
+    }
+}
```

Comment:
Quite cumbersome. A more straightforward way to implement it could be:
if (x < 0) x =0;
if (x >= w) x = w - 1;
if (y < 0) y = 0;
if (y >= h) y = h - 1;
return in_values[y * w + x];